File: oneEditAway.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <cmath>
using namespace std;
/*
字符串有三种编辑操作:插入一个英文字符、删除一个英文字符或者替换一个英文字符。
给定两个字符串，编写一个函数判定它们是否只需要一次(或者零次)编辑。
*/
// 双指针
bool oneEditAway(string first, string second)
{
    int len1 = first.size();
    int len2 = second.size();
    if ((first == second) || ((len1 == 0 && len2 == 1) || (len1 == 1 && len2 == 0)))
        return true;
    if (abs(len1 - len2) <= 1)
    {
        int a = 0, b = len1 - 1;
        int x = 0, y = len2 - 1;
        while (a <= b && a <= y && first[a] == second[b])
        {
            a++;
            x++;
        }
        while (b >= 0 && y >= 0 && first[b] == second[y])
        {
            b--;
            y--;
        }
        if (len1 > len2)
            return a == b;
        else
            return x == y;
    }
    return false;
}

// 官方题解，写法更优雅
bool oneEditAway(string first, string second)
{
    int m = first.size(), n = second.size();
    if (n - m == 1)
    {
        return oneInsert(first, second);
    }
    else if (m - n == 1)
    {
        return oneInsert(second, first);
    }
    else if (m == n)
    {
        bool foundDifference = false;
        for (int i = 0; i < m; i++)
        {
            if (first[i] != second[i])
            {
                if (!foundDifference)
                {
                    foundDifference = true;
                }
                else
                {
                    return false;
                }
            }
        }
        return true;
    }
    else
    {
        return false;
    }
}

bool oneInsert(string shorter, string longer)
{
    int length1 = shorter.size(), length2 = longer.size();
    int index1 = 0, index2 = 0;
    while (index1 < length1 && index2 < length2)
    {
        if (shorter[index1] == longer[index2])
        {
            index1++;
        }
        index2++;
        if (index2 - index1 > 1)
        {
            return false;
        }
    }
    return true;
}

File: oneEditAway.py
import math
"""
字符串有三种编辑操作:插入一个英文字符、删除一个英文字符或者替换一个英文字符。
给定两个字符串，编写一个函数判定它们是否只需要一次(或者零次)编辑。
"""
#双指针思路
def oneEditAway(self, first: str, second: str) -> bool:
    len1 = len(first)
    len2 = len(second)
    if first==second or ((len1==0 and len2==1) or (len1==1 and len2==0)):
        return True
    if abs(len1-len2)<=1:
        a=x=0      #a,b,x,y = 0,0,len1-1,len2-1
        b=len1-1
        y=len2-1
        while a<=b and x<=y and first[a]==second[x]:
            a+=1
            x+=1
        while b>=0 and y>=0 and first[b]==second[y]:
            b-=1
            y-=1
        if len1>len2:
            return a==b
        else:
            return x==y
    return False


#官方题解 更简洁
def oneEditAway(self, first: str, second: str) -> bool:
        m, n = len(first), len(second)
        if m < n:
            return self.oneEditAway(second, first)
        if m - n > 1:
            return False
        for i, (x, y) in enumerate(zip(first, second)):
            if x != y:
                return first[i + 1:] == second[i + 1:] if m == n else first[i + 1:] == second[i:]  # 注：改用下标枚举可达到 O(1) 空间复杂度
        return True
    

